### ISSUE TITLE
Max call stack size exceeded when using Flask

### ISSUE THREAD

**Issue by bschnurr** [microsoft/pylance-release#1211]
<h3> Environment data </h3>

-   Language Server version: 2021.4.3
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: false
-   python.analysis.typeCheckingMode: off

<h3> Expected behaviour </h3>

XXX

<h3> Actual behaviour </h3>

```
[Error - 3:07:38 PM] Request completionItem/resolve failed.
  Message: Request completionItem/resolve failed with message: Maximum call stack size exceeded
  Code: -32603 
```

<h3> Logs </h3>

Python Language Server Log
```
[Info  - 3:37:41 PM] Pylance language server 2021.4.3 (pyright 320a0a2f) starting
[Info  - 3:37:41 PM] Server root directory: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist
[Info  - 3:37:42 PM] No configuration file found.
[Info  - 3:37:42 PM] No pyproject.toml file found.
[Info  - 3:37:42 PM] Setting pythonPath for service "fastapi": "c:\Users\bschnurr\source\repos\fastapi\venv\Scripts\python.exe"
[Warn  - 3:37:42 PM] stubPath c:\Users\bschnurr\source\repos\fastapi\typings is not a valid directory.
[Info  - 3:37:42 PM] Assuming Python version 3.9
[Info  - 3:37:42 PM] Assuming Python platform Windows
Search paths for c:\Users\bschnurr\source\repos\fastapi
  c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib
  c:\Users\bschnurr\source\repos\fastapi
  c:\Users\bschnurr\source\repos\fastapi\typings
  c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\...
  c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\bundled\stubs
  C:\Users\bschnurr\AppData\Local\Programs\Python\Python39\DLLs
  C:\Users\bschnurr\AppData\Local\Programs\Python\Python39\lib
  C:\Users\bschnurr\AppData\Local\Programs\Python\Python39
  c:\Users\bschnurr\source\repos\fastapi\venv
  c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages
[Info  - 3:37:42 PM] Searching for source files
[Info  - 3:37:42 PM] Auto-excluding c:\Users\bschnurr\source\repos\fastapi\venv
[Info  - 3:37:42 PM] Found 3 source files
[FG] parsing: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (55ms)
[FG] parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 6ms] (248ms)
[FG] binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\builtins.pyi (71ms)
[FG] binding: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (1ms)
[Info  - 3:37:43 PM] Background analysis(1) root directory: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist
[Info  - 3:37:43 PM] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: ensurePartialStubPackages
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: setFileOpened
Background analysis message: setFileOpened
Background analysis message: getDiagnosticsForRange
Background analysis message: getSemanticTokens full
IntelliCode model c:\Users\bschnurr\.vscode\extensions\visualstudioexptteam.vscodeintellicode-1.2.12\cache\E61945A9A512ED5E1A3EE3F1A2365B88F8FE_E4E9EADA96734F01970E616FAB2FAC19
[BG(1)] getSemanticTokens full at c:\Users\bschnurr\source\repos\fastapi\fastai_test.py ...
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\fastapi\fastai_test.py (35ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 6ms] (212ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\builtins.pyi (73ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\fastapi\fastai_test.py (0ms)
[BG(1)] getSemanticTokens full at c:\Users\bschnurr\source\repos\fastapi\fastai_test.py (329ms)
Background analysis message: getSemanticTokens full
[BG(1)] getSemanticTokens full at c:\Users\bschnurr\source\repos\fastapi\LoginManager.py ...
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (9ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (1ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\typing.pyi [fs read 4ms] (58ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\typing.pyi (18ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 2ms] (14ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (2ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\typing-extensions\typing_extensions.pyi [fs read 0ms] (4ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\typing-extensions\typing_extensions.pyi (2ms)
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\__init__.py [fs read 1ms] (2ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\__init__.py (1ms)
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\login_manager.py [fs read 0ms] (26ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\login_manager.py (7ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\__init__.pyi [fs read 0ms] (13ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\__init__.pyi (1ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\app.pyi [fs read 0ms] (15ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\app.pyi (7ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\helpers.pyi [fs read 1ms] (5ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\helpers.pyi (2ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\abc.pyi [fs read 0ms] (1ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\abc.pyi (0ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\wrappers.pyi [fs read 0ms] (8ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\wrappers.pyi (0ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Werkzeug\werkzeug\wrappers.pyi [fs read 1ms] (8ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Werkzeug\werkzeug\wrappers.pyi (6ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\types.pyi [fs read 0ms] (11ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\types.pyi (4ms)
[BG(1)] getSemanticTokens full at c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (286ms)
Background analysis message: analyze
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py ...
[BG(1)]   checking: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (2ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (2ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getSemanticTokens range
[BG(1)] getSemanticTokens range 0:0 - 13:14 at c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (41ms)
Background analysis message: markFilesDirty
Background analysis message: analyze
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py ...
[BG(1)]   checking: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (11ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\LoginManager.py (11ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\fastai_test.py ...
[BG(1)]   checking: c:\Users\bschnurr\source\repos\fastapi\fastai_test.py (0ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\fastai_test.py (0ms)
Background analysis message: resumeAnalysis
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\main.py ...
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\fastapi\main.py [fs read 0ms] (11ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\fastapi\main.py (2ms)
[BG(1)]   checking: c:\Users\bschnurr\source\repos\fastapi\main.py (7ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\main.py (20ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Loading ONNX runtime...
Loaded ONNX runtime. Creating IntelliCode session...
Created IntelliCode session.
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: setFileOpened
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] parsing: c:\Users\bschnurr\source\repos\fastapi\main.py (8ms)
[FG] binding: c:\Users\bschnurr\source\repos\fastapi\main.py (5ms)
Background analysis message: getSemanticTokens full
[BG(1)] getSemanticTokens full at c:\Users\bschnurr\source\repos\fastapi\main.py ...
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\fastapi\main.py (3ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\fastapi\main.py (1ms)
[BG(1)] getSemanticTokens full at c:\Users\bschnurr\source\repos\fastapi\main.py (9ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\main.py ...
[BG(1)]   checking: c:\Users\bschnurr\source\repos\fastapi\main.py (2ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\fastapi\main.py (2ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] completion at c:\Users\bschnurr\source\repos\fastapi\main.py:13:14 ...
[FG]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\typing.pyi [fs read 5ms] (60ms)
[FG]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\typing.pyi (12ms)
[FG]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 2ms] (11ms)
[FG]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (2ms)
[FG]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\typing-extensions\typing_extensions.pyi [fs read 1ms] (4ms)
[FG]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\typing-extensions\typing_extensions.pyi (3ms)
[FG]   parsing: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\__init__.py [fs read 0ms] (3ms)
[FG]   binding: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\__init__.py (1ms)
[FG]   parsing: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\login_manager.py [fs read 3ms] (24ms)
[FG]   binding: c:\Users\bschnurr\source\repos\fastapi\venv\lib\site-packages\flask_login\login_manager.py (7ms)
[FG]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\__init__.pyi [fs read 1ms] (10ms)
[FG]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\__init__.pyi (1ms)
[FG]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\app.pyi [fs read 0ms] (12ms)
[FG]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\app.pyi (4ms)
[FG]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\helpers.pyi [fs read 0ms] (1ms)
[FG]   binding: c:\Users\bschnurr\.vscode\extensions\ms-python.vscode-pylance-2021.4.3\dist\typeshed-fallback\stubs\Flask\flask\helpers.pyi (1ms)
[FG] completion at c:\Users\bschnurr\source\repos\fastapi\main.py:13:14 [found 58 items] (198ms)
Recommendations: __dict__, buffer, hourangle, wait, primary_key, seed, async_step_user, random, getdefaultencoding, transform
Time taken to get recommendations: 13 ms, Memory increase: 233 KB.
[FG] parsing: C:\Users\bschnurr\AppData\Local\Temp\pyright-9312-O2mtiaOG17GQ\builtins-9312-1l1JNyuQKHZ3-.py [fs read 3ms] (120ms)
[FG] binding: C:\Users\bschnurr\AppData\Local\Temp\pyright-9312-O2mtiaOG17GQ\builtins-9312-1l1JNyuQKHZ3-.py (39ms)
[Error - 3:38:23 PM] Request completionItem/resolve failed.
  Message: Request completionItem/resolve failed with message: Maximum call stack size exceeded
  Code: -32603 

```

<h3> Code Snippet / Additional information </h3>

```python
from flask_login import LoginManager
from flask import Flask
app = Flask(__name__)

@app.route('/')
def hello_world():
   return 'Hello World'

if __name__ == '__main__':
   app.run()

login_manager = LoginManager()

login_manager.
```

**Comment by erictraut** [comment]
I'm not able to repro this in either pyright or pylance. Is the code snippet above all that is needed?

**Comment by bschnurr** [comment]
sorry. need to request a completion on 'login_manager'

**Comment by bschnurr** [comment]
update logs

**Comment by erictraut** [comment]
Yeah, I assumed that you needed to request a completion on `login_manager`. I tried that and wasn't able to repro. Maybe it's specific to the version of flask or flask_login that you have installed?

**Comment by bschnurr** [comment]
blis==0.7.4
catalogue==2.0.4
certifi==2020.12.5
chardet==4.0.0
click==7.1.2
cycler==0.10.0
cymem==2.0.5
fastai==2.1.10
fastcore==1.3.19
fastprogress==1.0.0
Flask==1.1.2
Flask-Login==0.5.0
idna==2.10
itsdangerous==1.1.0
Jinja2==2.11.3
joblib==1.0.1
kiwisolver==1.3.1
MarkupSafe==1.1.1
matplotlib==3.4.1
murmurhash==1.0.5
numpy==1.20.2
packaging==20.9
pandas==1.2.4
pathy==0.5.2
Pillow==8.2.0
preshed==3.0.5
pydantic==1.7.3
pyparsing==2.4.7
python-dateutil==2.8.1
pytz==2021.1
PyYAML==5.4.1
requests==2.25.1
scikit-learn==0.24.2
scipy==1.6.3
six==1.15.0
smart-open==3.0.0
spacy==3.0.6
spacy-legacy==3.0.5
srsly==2.4.1
thinc==8.0.3
threadpoolctl==2.1.0
torch==1.8.1
torchvision==0.9.1
tqdm==4.60.0
typer==0.3.2
typing-extensions==3.7.4.3
urllib3==1.26.4
wasabi==0.8.2
Werkzeug==1.0.1

**Comment by bschnurr** [comment]
Python 3.9.2

**Comment by bschnurr** [comment]
sorry i had additional packages installed not related. this is probably the minimum needed
Package      Version
------------ -------
click        7.1.2
Flask        1.1.2
Flask-Login  0.5.0
itsdangerous 1.1.0
Jinja2       2.11.3
MarkupSafe   1.1.1
pip          20.2.3
setuptools   49.2.1
Werkzeug     1.0.1

**Comment by bschnurr** [comment]
finally got a callstack.
```
RangeError: Maximum call stack size exceeded
    at new Map (<anonymous>)
    at new TypeVarMap (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:35270:28)
    at buildTypeVarMap (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:34248:24)
    at buildTypeVarMapFromSpecializedClass (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:34240:24)
    at partiallySpecializeType (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:33828:24)
    at getClassIterator (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:34052:41)
    at getClassIterator.next (<anonymous>)
    at getClassMemberIterator (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:33965:50)
    at getClassMemberIterator.next (<anonymous>)
    at Object.lookUpClassMember (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:33940:28)
```

**Comment by bschnurr** [comment]
might be a false callstack.
had some local changes. and now it says
```
RangeError: Maximum call stack size exceeded
    at String.replace (<anonymous>)
    at normalizeSlashes (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:40864:23)
    at Object.getPathComponents (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:40771:28)
    at PylanceImportResolver._splitPath (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:7471:44)
    at PylanceImportResolver.dirExistsCached (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:7395:32)
    at PylanceImportResolver.getSourceFilesFromStub (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:7168:22)
    at PylanceImportResolver.getSourceFilesFromStub (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\server.bundle.js:6476:29)
    at SourceMapper._isStubFile (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:14869:14)
    at d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:14562:26
    at SourceMapper._findMemberDeclarationsByName (d:\dev\bschnurr\pyrx\packages\vscode-pylance\dist\pyright.bundle.js:14548:21)
```

**Comment by bschnurr** [comment]
now I believe its due to _findFieldDeclarationsByName(..) not adding the uniqueId to the recursive map check

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.5.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202150-5-may-2021
